{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import torch\n",
    "\n",
    "import umap\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import sklearn.decomposition\n",
    "import sklearn.linear_model\n",
    "import sklearn.metrics\n",
    "\n",
    "import os\n",
    "import importlib\n",
    "import adjustText\n",
    "import seaborn as sns\n",
    "import tqdm.notebook as tqdm\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches\n",
    "\n",
    "import src.selector\n",
    "importlib.reload(selector)\n",
    "\n",
    "def getStanfordProfile(stanford_fold: str) -> pd.DataFrame:\n",
    "    # dataset\n",
    "    profile = pd.read_csv(os.path.join(stanford_fold, \"profile.txt\"))\n",
    "    # Isolate string (i.e. su001) -> int (i.e. 1)\n",
    "    profile[\"Isolate\"] = profile[\"Isolate\"].apply(lambda x: int(x[2:]))\n",
    "    # Treatment pre/post -> 0/1\n",
    "    profile[\"Treatment\"] = profile[\"Treatment\"].apply(lambda x: int(not \"pre\" in x))\n",
    "    # Sort by Isolate (1 to 8), Treatment (pre to post), and Tissue (normal to BCC)\n",
    "    profile = profile.sort_values(\n",
    "        by=[\"Isolate\", \"Treatment\", \"Tissue\"], ascending=[True, True, False]\n",
    "    ).reset_index(drop=True)\n",
    "    # only keep run, isolate, treatment, tissue\n",
    "    profile = profile[[\"Run\", \"Isolate\", \"Treatment\", \"Tissue\"]]\n",
    "    return profile\n",
    "\n",
    "def pearsonCorrelation(x1: np.ndarray, x2: np.ndarray) -> np.ndarray:\n",
    "    x1centered = x1 - x1.mean(axis=1, keepdims=True)        # (N1, 768)\n",
    "    x2centered = x2 - x2.mean(axis=1, keepdims=True)        # (N2, 768)\n",
    "    numerator = np.dot(x1centered, x2centered.T)            # (N1, N2)\n",
    "    x1var = np.sum(x1centered**2, axis=1, keepdims=True)    # (N1,  1)\n",
    "    x2var = np.sum(x2centered**2, axis=1, keepdims=True)    # (N2,  1)\n",
    "    denominator = np.sqrt(np.dot(x1var, x2var.T))           # (N1, N2)\n",
    "    return numerator / denominator                          # (N1, N2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# path\n",
    "stanford_fold = \"data/stanford/\"\n",
    "tcgaskcm_fold = \"data/tcga-skcm/\"\n",
    "feature_fold  = \"data/feature\"\n",
    "method = \"pretrain\"     # pretrain or finetune\n",
    "# parameter\n",
    "pval_thresh = 1e-5      \n",
    "ascending = False       # False for top max distance, True for top min distance\n",
    "top_k = 0.1\n",
    "pos_range = 32          # 32 for stanford, 18 for tcga-skcm\n",
    "mod = \"sep\"             # all or sep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" embd_stanford and mean \"\"\"\n",
    "\n",
    "profile_stanford = getStanfordProfile(stanford_fold)\n",
    "chr_list = [str(i) for i in range(1, 23)] + [\"X\"]   # BAM naming convention\n",
    "\n",
    "embd_stanford = {run: None for run in profile_stanford[\"Run\"]}\n",
    "for chromosome in chr_list:\n",
    "    feature_fold_chr = os.path.join(feature_fold, chromosome)\n",
    "    selection = selector.Selector(feature_fold_chr, ascending)\n",
    "    for run in profile_stanford[\"Run\"]:\n",
    "        embd_path = os.path.join(stanford_fold, f\"embd-{method}/{run}/{chromosome}.npy\")\n",
    "        embd_temp = selection.apply(\n",
    "            embd_path=embd_path, top_k=top_k, pos_range=pos_range, mod=mod\n",
    "        )\n",
    "        if embd_stanford[run] is None: \n",
    "            embd_stanford[run] = embd_temp\n",
    "        else:\n",
    "            embd_stanford[run] = np.concatenate((embd_stanford[run], embd_temp), axis=0)\n",
    "# combine all embd, remove all the nan row\n",
    "embd_all = np.concatenate([embd_stanford[run] for run in profile_stanford[\"Run\"]], axis=0)\n",
    "embd_all = embd_all[~np.isnan(embd_all).all(axis=1)]\n",
    "# calculate the mean and fill the embd nan row with mean\n",
    "# so that these nan will be 0 in pca\n",
    "mean = np.mean(embd_all, axis=0)\n",
    "for run in profile_stanford[\"Run\"]:\n",
    "    embd_stanford[run] = np.where(np.isnan(embd_stanford[run]), mean, embd_stanford[run])\n",
    "\n",
    "\"\"\" embd_tcgaskcm \"\"\"\n",
    "\n",
    "profile_patient = pd.read_csv(\"data/tcga-skcm/patient.csv\")\n",
    "profile_patient = profile_patient[['TCGA Participant Barcode', 'Leukocyte Fraction']]\n",
    "profile_sample = pd.read_csv(\"data/tcga-skcm/profile.tsv\", sep='\\t')\n",
    "profile_sample = profile_sample[['Case ID', 'Sample ID', 'Sample Type']]\n",
    "\n",
    "# keep only index 0 to 246 of sample\n",
    "profile_sample = profile_sample.iloc[:246]\n",
    "# keep only patient that sample in the profile_sample\n",
    "profile_patient = profile_patient[\n",
    "    profile_patient['TCGA Participant Barcode'].isin(profile_sample['Case ID'])\n",
    "]\n",
    "\n",
    "## get patients\n",
    "# sort by Leukocyte Fraction, keep the dataframe top 5 and bottom 5 patients\n",
    "profile_patient = profile_patient.sort_values(\n",
    "    by='Leukocyte Fraction', ascending=False\n",
    ")\n",
    "profile_patient = pd.concat(\n",
    "    [profile_patient.head(6), profile_patient.tail(6)], ignore_index=True\n",
    ")\n",
    "\n",
    "## get samples\n",
    "# filter the profile sample such that only keep the samples of the selected patients\n",
    "profile_sample = profile_sample[\n",
    "    profile_sample['Case ID'].isin(profile_patient['TCGA Participant Barcode'])\n",
    "]\n",
    "# set the patient's Leukocyte Fraction to all samples\n",
    "profile_sample['Leukocyte Fraction'] = profile_sample['Case ID'].apply(lambda x: profile_patient[profile_patient['TCGA Participant Barcode'] == x].iloc[0]['Leukocyte Fraction'])\n",
    "# sort by Leukocyte Fraction, then sample type\n",
    "profile_sample = profile_sample.sort_values(by=['Leukocyte Fraction', 'Sample Type'], ascending=[False, False])\n",
    "\n",
    "labels_tcgaskcm = []\n",
    "for run in profile_sample.index:\n",
    "    patient, sample, tissue, score = profile_sample.loc[run]\n",
    "    if tissue.startswith(\"Blood\"): tissue = \"NormalBlood\"\n",
    "    if tissue.startswith(\"Solid\"): tissue = \"NormalTissue\"\n",
    "    labels_tcgaskcm.append(f\"{patient};{tissue};{score:.3f}\")\n",
    "\n",
    "# if Leukocyte Fraction > 0.5, set to 1, else 0\n",
    "profile_sample['Leukocyte Fraction'] = profile_sample['Leukocyte Fraction'].apply(lambda x: 1 if x > 0.5 else 0)\n",
    "\n",
    "embd_tcgaskcm = {run: None for run in profile_sample.index}\n",
    "for chromosome in chr_list:\n",
    "    feature_fold_chr = os.path.join(feature_fold, chromosome)\n",
    "    selection = selector.Selector(feature_fold_chr, ascending)\n",
    "    for run in profile_sample.index:\n",
    "        embd_path = os.path.join(tcgaskcm_fold, f\"embd-{method}/{run:03d}/{chromosome}.npy\")\n",
    "        embd_temp = selection.apply(\n",
    "            embd_path=embd_path, top_k=top_k, pos_range=pos_range, mod=mod\n",
    "        )\n",
    "        if embd_tcgaskcm[run] is None: embd_tcgaskcm[run] = embd_temp\n",
    "        else: embd_tcgaskcm[run] = np.concatenate((embd_tcgaskcm[run], embd_temp), axis=0)\n",
    "for run in profile_sample.index:\n",
    "    embd_tcgaskcm[run] = np.where(np.isnan(embd_tcgaskcm[run]), mean, embd_tcgaskcm[run])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" pca_reducer_1: pca for len(runs) * [N,768] -> [len(runs), N] \"\"\"\n",
    "\n",
    "pca_reducer_1 = sklearn.decomposition.PCA()\n",
    "pca_reducer_1.fit(embd_all)\n",
    "print(\"pca_reducer_1\", pca_reducer_1.explained_variance_ratio_[:5])\n",
    "\n",
    "pca_1_stanford = np.array([pca_reducer_1.transform(embd_stanford[run]) for run in profile_stanford[\"Run\"]])[:, :, 0]\n",
    "pca_1_tcgaskcm = np.array([pca_reducer_1.transform(embd_tcgaskcm[run]) for run in profile_sample.index])[:, :, 0]\n",
    "\n",
    "\"\"\" pca_reducer_2: pca for [len(runs), N] -> [len(runs), 10] \"\"\"\n",
    "\n",
    "pca_reducer_2 = sklearn.decomposition.PCA(n_components=10)\n",
    "pca_2_stanford = pca_reducer_2.fit_transform(pca_1_stanford)\n",
    "print(\"pca_reducer_2\", pca_reducer_2.explained_variance_ratio_[:5])\n",
    "\n",
    "pca_reducer_2 = sklearn.decomposition.PCA(n_components=10)\n",
    "pca_2_tcgaskcm = pca_reducer_2.fit_transform(pca_1_tcgaskcm)\n",
    "print(\"pca_reducer_2\", pca_reducer_2.explained_variance_ratio_[:5])\n",
    "\n",
    "\"\"\" lr: logistic regression using pca_1 [len(runs), N] \"\"\"\n",
    "\n",
    "lr = sklearn.linear_model.LogisticRegression()\n",
    "t_stanford = profile_stanford[\"Treatment\"].to_numpy()\n",
    "t_tcgaskcm = profile_sample[\"Leukocyte Fraction\"].to_numpy()\n",
    "lr.fit(pca_1_stanford, t_stanford)\n",
    "y_stanford = lr.predict(pca_1_stanford)\n",
    "y_tcgaskcm = lr.predict(pca_1_tcgaskcm)\n",
    "print(sklearn.metrics.confusion_matrix(t_stanford, y_stanford))\n",
    "print(sklearn.metrics.confusion_matrix(t_tcgaskcm, y_tcgaskcm))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = pca_2_stanford[:, :4]\n",
    "print(temp.shape)\n",
    "profile_stanford[\"PCA1\"] = temp[:, 0]\n",
    "profile_stanford[\"PCA2\"] = temp[:, 1]\n",
    "profile_stanford[\"PCA3\"] = temp[:, 2]\n",
    "profile_stanford[\"PCA4\"] = temp[:, 3]\n",
    "profile_stanford.to_csv(\"data/stanford/pca.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = pca_2_tcgaskcm[:, :4]\n",
    "\n",
    "profile_patient = pd.read_csv(\"data/tcga-skcm/patient.csv\")\n",
    "profile_patient = profile_patient[['TCGA Participant Barcode', 'Leukocyte Fraction']]\n",
    "profile_sample = pd.read_csv(\"data/tcga-skcm/profile.tsv\", sep='\\t')\n",
    "profile_sample = profile_sample[['Case ID', 'Sample ID', 'Sample Type']]\n",
    "\n",
    "# keep only index 0 to 246 of sample\n",
    "profile_sample = profile_sample.iloc[:246]\n",
    "# keep only patient that sample in the profile_sample\n",
    "profile_patient = profile_patient[\n",
    "    profile_patient['TCGA Participant Barcode'].isin(profile_sample['Case ID'])\n",
    "]\n",
    "\n",
    "## get patients\n",
    "# sort by Leukocyte Fraction, keep the dataframe top 5 and bottom 5 patients\n",
    "profile_patient = profile_patient.sort_values(\n",
    "    by='Leukocyte Fraction', ascending=False\n",
    ")\n",
    "profile_patient = pd.concat(\n",
    "    [profile_patient.head(6), profile_patient.tail(6)], ignore_index=True\n",
    ")\n",
    "\n",
    "## get samples\n",
    "# filter the profile sample such that only keep the samples of the selected patients\n",
    "profile_sample = profile_sample[\n",
    "    profile_sample['Case ID'].isin(profile_patient['TCGA Participant Barcode'])\n",
    "]\n",
    "# set the patient's Leukocyte Fraction to all samples\n",
    "profile_sample['Leukocyte Fraction'] = profile_sample['Case ID'].apply(lambda x: profile_patient[profile_patient['TCGA Participant Barcode'] == x].iloc[0]['Leukocyte Fraction'])\n",
    "# sort by Leukocyte Fraction, then sample type\n",
    "profile_sample = profile_sample.sort_values(by=['Leukocyte Fraction', 'Sample Type'], ascending=[False, False])\n",
    "# if Leukocyte Fraction > 0.5, set to 1, else 0\n",
    "#profile_sample['Leukocyte Fraction'] = profile_sample['Leukocyte Fraction'].apply(lambda x: 1 if x > 0.5 else 0)\n",
    "\n",
    "profile_sample[\"PCA1\"] = temp[:, 0]\n",
    "profile_sample[\"PCA2\"] = temp[:, 1]\n",
    "profile_sample[\"PCA3\"] = temp[:, 2]\n",
    "profile_sample[\"PCA4\"] = temp[:, 3]\n",
    "profile_sample[\"Prediction\"] = y_tcgaskcm\n",
    "profile_sample.to_csv(\"data/tcga-skcm/pca.csv\", index=False)\n",
    "\n",
    "print(profile_sample)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" plot scatter of PCA1 and PCA2 of all runs \"\"\"\n",
    "\n",
    "fontsize=6\n",
    "\n",
    "labels_stanford = []\n",
    "for run in profile_stanford[\"Run\"]:\n",
    "    id, isolate, treatment, tissue = profile_stanford[profile_stanford[\"Run\"] == run].iloc[0]\n",
    "    id = id[7:]\n",
    "    treatment = \"Pre\" if treatment == 0 else \"Post\"\n",
    "    tissue = \"Normal\" if \"normal\" in tissue else \"Tumor\"\n",
    "    labels_stanford.append(f\"{isolate};{id};{treatment};{tissue}\")\n",
    "\n",
    "# color by isolate\n",
    "plt.figure(figsize=(5, 5), dpi=300)\n",
    "colors = [f\"C{i}\" for i in range(8)]\n",
    "plt.scatter(pca_2_stanford[:, 0], pca_2_stanford[:, 1], c=[colors[profile_stanford[profile_stanford[\"Run\"] == run].iloc[0][\"Isolate\"]-1] for run in profile_stanford[\"Run\"]])\n",
    "for i in range(len(profile_stanford)):\n",
    "    plt.text(pca_2_stanford[i, 0]-7, pca_2_stanford[i, 1]+4, labels_stanford[i], fontsize=fontsize)\n",
    "plt.legend(\n",
    "    handles=[\n",
    "        matplotlib.patches.Patch(color=colors[i], label=f\"patient {i+1}\") for i in range(8)\n",
    "    ], loc=\"lower left\", ncol=2, fontsize=fontsize+0.5\n",
    ")\n",
    "plt.xlim(-50, 25)\n",
    "plt.ylim(-100, 90)\n",
    "plt.show()\n",
    "# color by treatment\n",
    "plt.figure(figsize=(5, 5), dpi=300)\n",
    "colors = [\"red\" if profile_stanford[profile_stanford[\"Run\"] == run].iloc[0][\"Treatment\"] == 0 else \"blue\" for run in profile_stanford[\"Run\"]]\n",
    "plt.scatter(pca_2_stanford[:, 0], pca_2_stanford[:, 1], c=colors)\n",
    "for i in range(len(profile_stanford)):\n",
    "    plt.text(pca_2_stanford[i, 0]-7, pca_2_stanford[i, 1]+4, labels_stanford[i], fontsize=fontsize)\n",
    "plt.legend(\n",
    "    handles=[\n",
    "        matplotlib.patches.Patch(color=\"r\", label=\"pre anti-PD1 treatment\"), \n",
    "        matplotlib.patches.Patch(color=\"b\", label=\"post anti-PD1 treatment\")\n",
    "    ], loc=\"lower left\", ncol=1, fontsize=fontsize+0.5\n",
    ")\n",
    "plt.xlim(-50, 25)\n",
    "plt.ylim(-100, 90)\n",
    "plt.show()\n",
    "# color by tissue\n",
    "plt.figure(figsize=(5, 5), dpi=300)\n",
    "colors = [\"green\" if \"normal\" in profile_stanford[profile_stanford[\"Run\"] == run].iloc[0][\"Tissue\"] else \"purple\" for run in profile_stanford[\"Run\"]]\n",
    "plt.scatter(pca_2_stanford[:, 0], pca_2_stanford[:, 1], c=colors)\n",
    "for i in range(len(profile_stanford)):\n",
    "    plt.text(pca_2_stanford[i, 0]-7, pca_2_stanford[i, 1]+4, labels_stanford[i], fontsize=fontsize)\n",
    "plt.legend(\n",
    "    handles=[\n",
    "        matplotlib.patches.Patch(color=\"g\", label=\"normal skin\"), \n",
    "        matplotlib.patches.Patch(color=\"purple\", label=\"BCC tumor\")\n",
    "    ], loc=\"lower left\", ncol=1, fontsize=fontsize+0.5\n",
    ")\n",
    "plt.xlim(-50, 25)\n",
    "plt.ylim(-100, 90)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" plot correlation heat map of runs for selected pos\"\"\"\n",
    "\n",
    "# TODO: clusting of heat map\n",
    "\n",
    "runs = profile_stanford[\"Run\"][\"Run\"].to_list()\n",
    "\n",
    "# calculate the correlation between each pair of runs\n",
    "heat_map = pearsonCorrelation(pca_1, pca_1)     # [len(runs), len(runs)]\n",
    "for i in range(len(heat_map)): heat_map[i, i] = np.nan\n",
    "\n",
    "# plot the heat map 6*6 \n",
    "labels = []\n",
    "for run in runs:\n",
    "    id, isolate, treatment, tissue = profile_stanford[\"Run\"][profile_stanford[\"Run\"][\"Run\"] == run].iloc[0]\n",
    "    labels.append(\"{}-{}-{}-{}\".format(\n",
    "            id, isolate,\n",
    "            \"Pre\" if treatment == 0 else \"Post\", \n",
    "            \"Normal\" if \"normal\" in tissue else \"Tumor\"\n",
    "        ))\n",
    "plt.figure(dpi=300, figsize=(20, 18))\n",
    "sns.heatmap(\n",
    "    heat_map, annot=True, fmt=\".3f\", cmap=\"coolwarm\", square=True,\n",
    "    yticklabels=labels, xticklabels=False\n",
    ")\n",
    "plt.title(\"Pearson Correlation\", fontweight='bold')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" plot scatter of PCA1 and PCA2 of all runs \"\"\"\n",
    "\n",
    "fontsize=6\n",
    "\n",
    "# color by patient\n",
    "colors = [\"C1\", \"C2\", \"C2\", \"C3\", \"C4\", \"C4\", \"C5\", \"C6\", \"C7\", \"C7\", \"C8\", \"C9\", \"C9\", \"C10\", \"C11\", \"C11\", \"C12\"]\n",
    "plt.figure(figsize=(5, 5), dpi=300)\n",
    "plt.scatter(pca_2_tcgaskcm[:, 0], pca_2_tcgaskcm[:, 1], c=[colors[i] for i in range(len(profile_sample))])\n",
    "for i in range(len(labels_tcgaskcm)):\n",
    "    plt.text(pca_2_tcgaskcm[i, 0]-0.5, pca_2_tcgaskcm[i, 1]+0.05, labels_tcgaskcm[i], fontsize=fontsize)\n",
    "plt.legend(\n",
    "    handles=[\n",
    "        matplotlib.patches.Patch(color=\"C1\", label=\"TCGA-EE-A2GK\"), \n",
    "        matplotlib.patches.Patch(color=\"C2\", label=\"TCGA-EB-A5VV\"),\n",
    "        matplotlib.patches.Patch(color=\"C3\", label=\"TCGA-EE-A2ME\"),\n",
    "        matplotlib.patches.Patch(color=\"C4\", label=\"TCGA-FR-A7UA\"),\n",
    "        matplotlib.patches.Patch(color=\"C5\", label=\"TCGA-FW-A5DY\"),\n",
    "        matplotlib.patches.Patch(color=\"C6\", label=\"TCGA-EE-A3JD\"),\n",
    "        matplotlib.patches.Patch(color=\"C7\", label=\"TCGA-D3-A8GP\"),\n",
    "        matplotlib.patches.Patch(color=\"C8\", label=\"TCGA-D9-A6EC\"),\n",
    "        matplotlib.patches.Patch(color=\"C9\", label=\"TCGA-EE-A3AD\"),\n",
    "        matplotlib.patches.Patch(color=\"C10\", label=\"TCGA-EE-A3AH\"),\n",
    "        matplotlib.patches.Patch(color=\"C11\", label=\"TCGA-D3-A3ML\"),\n",
    "        matplotlib.patches.Patch(color=\"C12\", label=\"TCGA-FS-A1ZY\"),\n",
    "    ], loc=\"lower right\", ncol=2, fontsize=fontsize+0.5\n",
    ")\n",
    "plt.show()\n",
    "# color by ground truth\n",
    "plt.figure(figsize=(5, 5), dpi=300)\n",
    "colors = [\"red\" if profile_sample.loc[run]['Leukocyte Fraction'] == 0 else \"blue\" for run in profile_sample.index]\n",
    "plt.scatter(pca_2_tcgaskcm[:, 0], pca_2_tcgaskcm[:, 1], c=colors)\n",
    "texts = [\n",
    "    plt.text(pca_2_tcgaskcm[i, 0], pca_2_tcgaskcm[i, 1], labels_tcgaskcm[i], fontsize=fontsize) \n",
    "    for i in range(len(labels_tcgaskcm))\n",
    "]\n",
    "adjustText.adjust_text(texts)\n",
    "plt.legend(\n",
    "    handles=[\n",
    "        matplotlib.patches.Patch(color=\"r\", label=\"Leukocyte Fraction < 0.5\"), \n",
    "        matplotlib.patches.Patch(color=\"b\", label=\"Leukocyte Fraction > 0.5\")\n",
    "    ], loc=\"lower right\", ncol=1, fontsize=fontsize+0.5\n",
    ")\n",
    "plt.show()\n",
    "# color by prediction\n",
    "plt.figure(figsize=(5, 5), dpi=300)\n",
    "colors = [\"red\" if y_tcgaskcm[i] == 0 else \"blue\" for i in range(len(y_tcgaskcm))]\n",
    "plt.scatter(pca_2_tcgaskcm[:, 0], pca_2_tcgaskcm[:, 1], c=colors)\n",
    "for i in range(len(labels_tcgaskcm)):\n",
    "    plt.text(pca_2_tcgaskcm[i, 0]-0.5, pca_2_tcgaskcm[i, 1]+0.05, labels_tcgaskcm[i], fontsize=fontsize)\n",
    "plt.legend(\n",
    "    handles=[\n",
    "        matplotlib.patches.Patch(color=\"r\", label=\"Leukocyte Fraction < 0.5\"), \n",
    "        matplotlib.patches.Patch(color=\"b\", label=\"Leukocyte Fraction > 0.5\")\n",
    "    ], loc=\"lower right\", ncol=1, fontsize=fontsize+0.5\n",
    ")\n",
    "plt.show()\n",
    "# color by tissue\n",
    "plt.figure(figsize=(5, 5), dpi=300)\n",
    "colors = []\n",
    "for run in profile_sample.index:\n",
    "    if profile_sample.loc[run]['Sample Type'].startswith(\"Blood\"): colors.append(\"green\")\n",
    "    elif profile_sample.loc[run]['Sample Type'].startswith(\"Metastatic\"): colors.append(\"purple\")\n",
    "    else: colors.append(\"blue\")\n",
    "plt.scatter(pca_2_tcgaskcm[:, 0], pca_2_tcgaskcm[:, 1], c=colors)\n",
    "for i in range(len(labels_tcgaskcm)):\n",
    "    plt.text(pca_2_tcgaskcm[i, 0]-0.5, pca_2_tcgaskcm[i, 1]+0.05, labels_tcgaskcm[i], fontsize=fontsize)\n",
    "plt.legend(\n",
    "    handles=[\n",
    "        matplotlib.patches.Patch(color=\"green\", label=\"NormalBlood\"), \n",
    "        matplotlib.patches.Patch(color=\"purple\", label=\"Metastatic\"),\n",
    "        matplotlib.patches.Patch(color=\"blue\", label=\"NormalTissue\"),\n",
    "    ], loc=\"lower right\", ncol=1, fontsize=fontsize+0.5\n",
    ")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" Plot Hexbin N * N+1\"\"\"\n",
    "\n",
    "\"\"\"\n",
    "gridsize = 80\n",
    "vmin, vmax = 4, 64\n",
    "fontsize = 18\n",
    "\n",
    "fig, axs = plt.subplots(\n",
    "    len(runs), len(runs)+1, figsize=(35, 30), \n",
    "    sharex=True, sharey=True\n",
    ")\n",
    "# plot\n",
    "for i in range(len(runs)):\n",
    "    # plot umpa of sample i before filter in diagonal\n",
    "    axs[i, i].hexbin(\n",
    "        umap[runs[i]][:, 0], umap[runs[i]][:, 1], \n",
    "        cmap=\"Reds\", gridsize=gridsize, vmin=vmin, vmax=vmax,\n",
    "    )\n",
    "    # plot umap of (i, :) and (:, i) after filter by distance (i, j)\n",
    "    for j in range(i+1, len(runs)):\n",
    "        # umap of (i, :)\n",
    "        axs[i, j].hexbin(\n",
    "            umap[runs[i]][:, 0], umap[runs[i]][:, 1], \n",
    "            cmap=\"Reds\", gridsize=gridsize, vmin=vmin, vmax=vmax/2,\n",
    "        )\n",
    "        # umap of (:, i)\n",
    "        axs[j, i].hexbin(\n",
    "            umap[runs[j]][:, 0], umap[runs[j]][:, 1], \n",
    "            cmap=\"Reds\", gridsize=gridsize, vmin=vmin, vmax=vmax/2,\n",
    "        )\n",
    "    # plot umap of (i, len(runs)) of combine all filter\n",
    "    axs[i, len(runs)].hexbin(\n",
    "        umap[runs[i]][:, 0], \n",
    "        umap[runs[i]][:, 1], \n",
    "        cmap=\"Reds\", gridsize=gridsize, vmin=vmin, vmax=vmax/2,\n",
    "    )\n",
    "# set up axis and label\n",
    "for i in range(len(runs)):\n",
    "    id, isolate, treatment, tissue = profile[profile[\"Run\"] == runs[i]].iloc[0]\n",
    "    treatment = \"Pre\" if treatment == 0 else \"Post\"\n",
    "    tissue = \"Normal\" if \"normal\" in tissue else \"Tumor\"\n",
    "    info = f\"{id} - {isolate} - {treatment} - {tissue}\"\n",
    "    axs[0,  i].set_xlabel(info, fontsize=fontsize, fontweight='bold')\n",
    "    axs[0,  i].xaxis.set_label_position(\"top\")\n",
    "    axs[0, len(runs)].set_xlabel(\"All Selected Reads\", fontsize=fontsize, fontweight='bold')\n",
    "    axs[0, len(runs)].xaxis.set_label_position(\"top\")\n",
    "    axs[i, -1].set_ylabel(info, fontsize=fontsize, fontweight='bold')\n",
    "    axs[i, -1].yaxis.set_label_position(\"right\")\n",
    "    for j in range(len(runs)+1):\n",
    "        axs[i, j].set_aspect(\"equal\")\n",
    "        axs[i, j].set_xlim(-22, 22)\n",
    "        axs[i, j].set_ylim(-22, 22)\n",
    "fig.supxlabel(\"UMAP1\", fontsize=fontsize, fontweight='bold', y=0.004)\n",
    "fig.supylabel(\"UMAP2\", fontsize=fontsize, fontweight='bold', x=0.008)\n",
    "fig.tight_layout()\n",
    "fig.savefig(\"temp.png\", dpi=500)\n",
    "fig.show()\n",
    "\"\"\"\n",
    "\n",
    "\"\"\" Plot Hexbin of all samples \"\"\"\n",
    "\n",
    "\"\"\"\n",
    "gridsize = 80\n",
    "vmin, vmax = 4, 64\n",
    "fontsize = 18\n",
    "\n",
    "# hexbin, split by isolate\n",
    "fig, axs = plt.subplots(1, 2, figsize=(10, 5), sharex=True, sharey=True)\n",
    "umap_isolates = [None, None]\n",
    "for i in range(len(runs)):\n",
    "    axs_index = 0 if profile[profile[\"Run\"] == runs[i]][\"Isolate\"].values[0] == 3 else 1\n",
    "    if umap_isolates[axs_index] is None:\n",
    "        umap_isolates[axs_index] = umap[runs[i]]\n",
    "    else:\n",
    "        umap_isolates[axs_index] = np.concatenate(\n",
    "            [umap_isolates[axs_index], umap[runs[i]]]\n",
    "        )\n",
    "for axs_index in range(2):\n",
    "    axs[axs_index].hexbin(\n",
    "        umap_isolates[axs_index][:, 0], umap_isolates[axs_index][:, 1], \n",
    "        cmap=\"Reds\", gridsize=gridsize, vmin=vmin, vmax=vmax,\n",
    "    )\n",
    "    axs[axs_index].set_xlim(-22, 22)\n",
    "    axs[axs_index].set_ylim(-22, 22)\n",
    "    axs[axs_index].set_title(\n",
    "        \"su003\" if axs_index == 0 else \"su007\", \n",
    "        fontweight='bold', fontsize=fontsize\n",
    "    )\n",
    "fig.tight_layout()\n",
    "fig.show()\n",
    "\n",
    "# hexbin, split by treatment\n",
    "fig, axs = plt.subplots(1, 2, figsize=(10, 5), sharex=True, sharey=True)\n",
    "umap_treatments = [None, None]\n",
    "for i in range(len(runs)):\n",
    "    axs_index = profile[profile[\"Run\"] == runs[i]][\"Treatment\"].values[0]\n",
    "    if umap_treatments[axs_index] is None:\n",
    "        umap_treatments[axs_index] = umap[runs[i]]\n",
    "    else:\n",
    "        umap_treatments[axs_index] = np.concatenate(\n",
    "            [umap_treatments[axs_index], umap[runs[i]]]\n",
    "        )\n",
    "for axs_index in range(2):\n",
    "    axs[axs_index].hexbin(\n",
    "        umap_treatments[axs_index][:, 0], umap_treatments[axs_index][:, 1], \n",
    "        cmap=\"Reds\", gridsize=gridsize,\n",
    "        vmin = vmin if axs_index == 0 else (vmin/2),\n",
    "        vmax = vmax if axs_index == 0 else (vmax/2)\n",
    "    )\n",
    "    axs[axs_index].set_xlim(-22, 22)\n",
    "    axs[axs_index].set_ylim(-22, 22)\n",
    "    axs[axs_index].set_title(\n",
    "        \"Pre\" if axs_index == 0 else \"Post\", \n",
    "        fontweight='bold', fontsize=fontsize\n",
    "    )\n",
    "fig.tight_layout()\n",
    "fig.show()\n",
    "\"\"\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "TCGA-Onc",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
