{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import umap\n",
    "import sklearn.decomposition\n",
    "import sklearn.svm, sklearn.linear_model\n",
    "import sklearn.metrics, sklearn.preprocessing\n",
    "\n",
    "import os\n",
    "import tqdm\n",
    "import pysam\n",
    "import subprocess\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from src import Selector\n",
    "\n",
    "selector = Selector(\"data/feature\")\n",
    "\n",
    "chromosome_list = [str(i) for i in range(1, 23)] + [\"X\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Sample Preparation\n",
    "\n",
    "import os, pysam\n",
    "\n",
    "def make_sample(bam_load_path, bam_save_path, num_reads=100):\n",
    "    # file check\n",
    "    os.makedirs(os.path.dirname(bam_save_path), exist_ok=True)\n",
    "    if os.path.exists(bam_save_path): os.remove(bam_save_path)\n",
    "    if os.path.exists(bam_save_path + \".bai\"): os.remove(bam_save_path + \".bai\")\n",
    "    # make sample\n",
    "    input = pysam.AlignmentFile(bam_load_path, \"rb\")\n",
    "    output = pysam.AlignmentFile(bam_save_path, \"wb\", template=input)\n",
    "    for c in [str(i) for i in range(1, 23)] + [\"X\"]:\n",
    "        count = 0\n",
    "        for read in input.fetch(c):\n",
    "            if count >= num_reads: break\n",
    "            output.write(read)\n",
    "            count += 1\n",
    "    input.close()\n",
    "    output.close()\n",
    "    _ = pysam.sort(\"-o\", bam_save_path, bam_save_path)\n",
    "    _ = pysam.index(bam_save_path)\n",
    "\n",
    "make_sample(\"/mnt/efs_v2/dbgap_tcga/users/tianrui.qi/SIP-DB2/data/bam/SRR8924580.bam\", \"temp/bam/a.bam\")\n",
    "make_sample(\"/mnt/efs_v2/dbgap_tcga/users/tianrui.qi/SIP-DB2/data/bam/SRR8924581.bam\", \"temp/bam/b.bam\")\n",
    "make_sample(\"/mnt/efs_v2/dbgap_tcga/users/tianrui.qi/SIP-DB2/data/bam/SRR8924582.bam\", \"temp/bam/c.bam\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of reads in temp/hdf/a.h5/chr2: 90\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>pos</th>\n",
       "      <th>1e+00</th>\n",
       "      <th>1e-01</th>\n",
       "      <th>1e-02</th>\n",
       "      <th>1e-03</th>\n",
       "      <th>1e-04</th>\n",
       "      <th>1e-05</th>\n",
       "      <th>1e-06</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CCGACCCCCGACCCCCACCCCGAACCCGACCCCGACCCCGACCCAA...</td>\n",
       "      <td>10475</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GACCCCCGACCCCCACCCCGAACCCGACCCCGACCCCGACCCAAAC...</td>\n",
       "      <td>10477</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CCCCACCCCCCACCCCCACCCCCAACCCGACCCCGACCCCGACCCA...</td>\n",
       "      <td>10481</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TAGCCCTAGCCCTAGCCCTAACCCCTATCCCCTAACCCTAAGCCGA...</td>\n",
       "      <td>10551</td>\n",
       "      <td>16.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CCCCTAACCCCTAACCCTAAGCCGAAGCCTAACTCGTGTCTGACTT...</td>\n",
       "      <td>10572</td>\n",
       "      <td>20.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            sequence    pos  1e+00  1e-01  \\\n",
       "0  CCGACCCCCGACCCCCACCCCGAACCCGACCCCGACCCCGACCCAA...  10475    5.0    0.0   \n",
       "1  GACCCCCGACCCCCACCCCGAACCCGACCCCGACCCCGACCCAAAC...  10477    5.0    0.0   \n",
       "2  CCCCACCCCCCACCCCCACCCCCAACCCGACCCCGACCCCGACCCA...  10481    6.0    0.0   \n",
       "3  TAGCCCTAGCCCTAGCCCTAACCCCTATCCCCTAACCCTAAGCCGA...  10551   16.0    4.0   \n",
       "4  CCCCTAACCCCTAACCCTAAGCCGAAGCCTAACTCGTGTCTGACTT...  10572   20.0    5.0   \n",
       "\n",
       "   1e-02  1e-03  1e-04  1e-05  1e-06  \n",
       "0    0.0    0.0    0.0    0.0    0.0  \n",
       "1    0.0    0.0    0.0    0.0    0.0  \n",
       "2    0.0    0.0    0.0    0.0    0.0  \n",
       "3    0.0    0.0    0.0    0.0    0.0  \n",
       "4    0.0    0.0    0.0    0.0    0.0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>pos</th>\n",
       "      <th>1e+00</th>\n",
       "      <th>1e-01</th>\n",
       "      <th>1e-02</th>\n",
       "      <th>1e-03</th>\n",
       "      <th>1e-04</th>\n",
       "      <th>1e-05</th>\n",
       "      <th>1e-06</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>TTTCTTTGCATCTCTGTGTCATATATCACCAAATCTGCCTAAGCTT...</td>\n",
       "      <td>16008</td>\n",
       "      <td>12.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>AACCTGATATGTAGTAGTTCCTCATGAAATACTAGCTCTGTTGATA...</td>\n",
       "      <td>16494</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>TCCTCATGAAATACTAGCTCTGTTGATAAAACTAGACTGAAAGGAG...</td>\n",
       "      <td>16512</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>CTCCTTTGTGCGCCTGGGGTTGCTGGCTGGCACTGGGCAGCAGGCA...</td>\n",
       "      <td>17267</td>\n",
       "      <td>6.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>TGCGCCTGGGGTTGCTGGCTGGCACTGGGCAGCAGGCAGCAGAGCT...</td>\n",
       "      <td>17275</td>\n",
       "      <td>6.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             sequence    pos  1e+00  1e-01  \\\n",
       "85  TTTCTTTGCATCTCTGTGTCATATATCACCAAATCTGCCTAAGCTT...  16008   12.0    2.0   \n",
       "86  AACCTGATATGTAGTAGTTCCTCATGAAATACTAGCTCTGTTGATA...  16494    2.0    1.0   \n",
       "87  TCCTCATGAAATACTAGCTCTGTTGATAAAACTAGACTGAAAGGAG...  16512    2.0    1.0   \n",
       "88  CTCCTTTGTGCGCCTGGGGTTGCTGGCTGGCACTGGGCAGCAGGCA...  17267    6.0    2.0   \n",
       "89  TGCGCCTGGGGTTGCTGGCTGGCACTGGGCAGCAGGCAGCAGAGCT...  17275    6.0    2.0   \n",
       "\n",
       "    1e-02  1e-03  1e-04  1e-05  1e-06  \n",
       "85    0.0    0.0    0.0    0.0    0.0  \n",
       "86    0.0    0.0    0.0    0.0    0.0  \n",
       "87    0.0    0.0    0.0    0.0    0.0  \n",
       "88    0.0    0.0    0.0    0.0    0.0  \n",
       "89    0.0    0.0    0.0    0.0    0.0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## BAM and SNPs to Sequence\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "# run command below in terminal\n",
    "# can be run in parallel since bam2seq designed to be sample level parallel\n",
    "# python main.py bam2seq -B temp/bam/a.bam -S /mnt/efs_v2/dbgap_tcga/users/tianrui.qi/SIP-DB2/data/snps.csv -H temp/hdf/a.h5\n",
    "# python main.py bam2seq -B temp/bam/b.bam -S /mnt/efs_v2/dbgap_tcga/users/tianrui.qi/SIP-DB2/data/snps.csv -H temp/hdf/b.h5\n",
    "# python main.py bam2seq -B temp/bam/c.bam -S /mnt/efs_v2/dbgap_tcga/users/tianrui.qi/SIP-DB2/data/snps.csv -H temp/hdf/c.h5\n",
    "\n",
    "# read hdf of given sample and chromosome\n",
    "hdf_path = \"temp/hdf/a.h5\"\n",
    "chromosome = \"2\"\n",
    "hdf = pd.read_hdf(hdf_path, key=f\"/chr{chromosome}\", mode='r')\n",
    "# filter reads that cover at least one variants with p-value<pval_thresh\n",
    "#pval_thresh = 1e-1\n",
    "#hdf = hdf[hdf[f\"{pval_thresh:.0e}\"]>=1]\n",
    "# glace\n",
    "print(f\"number of reads in {hdf_path}/chr{chromosome}: {len(hdf)}\")\n",
    "display(hdf.head())\n",
    "display(hdf.tail())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[16008. 16494. 16512.]\n",
      "[17267. 17275.]\n"
     ]
    }
   ],
   "source": [
    "## Sequence to Embedding\n",
    "\n",
    "# run command below in terminal\n",
    "# can be run in parallel since seq2embd designed to be sample level parallel\n",
    "# python main.py seq2embd -H temp/hdf/a.h5 -E temp/embd/a\n",
    "# python main.py seq2embd -H temp/hdf/b.h5 -E temp/embd/b\n",
    "# python main.py seq2embd -H temp/hdf/c.h5 -E temp/embd/c\n",
    "\n",
    "import numpy as np\n",
    "a = np.load(\"temp/embd/a/2/000/016.npy\")\n",
    "print(a[:, 768])\n",
    "a = np.load(\"temp/embd/a/2/000/017.npy\")\n",
    "print(a[:, 768])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "addFeature/2: 100%|██████████| 243000/243000 [00:08<00:00, 28053.02hash/s]\n",
      "addFeature/2: 100%|██████████| 243000/243000 [00:05<00:00, 46866.85hash/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "feature 2/000/010\n",
      "[1. 1. 2. 1. 0. 0. 0. 1. 1. 0. 0. 0. 1. 1. 1. 1. 0. 1.]\n",
      "[10207. 10270. 10287. 10349. 10475. 10477. 10481. 10409. 10470. 10551.\n",
      " 10572. 10587. 10520. 10545. 10589. 10627. 10720. 10705.]\n",
      "sample a/2/000/010\n",
      "[10475. 10477. 10481. 10551. 10572. 10587. 10720.]\n",
      "sample b/2/000/010\n",
      "[10207. 10270. 10349. 10409. 10470. 10520. 10545. 10589. 10627. 10705.]\n",
      "sample c/2/000/010\n",
      "[10287.]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "## Add Feature (sample_idx, pos, embd_idx, distance)\n",
    "\n",
    "import numpy as np\n",
    "import os\n",
    "import shutil\n",
    "from src.embd2repre import Selector\n",
    "\n",
    "if os.path.exists(\"temp/feature\"): shutil.rmtree(\"temp/feature\")\n",
    "\n",
    "selector = Selector(\"temp/feature\")\n",
    "selector.addFeature([\"temp/embd/a\", \"temp/embd/b\"], chromosome=\"2\")\n",
    "selector.addFeature([\"temp/embd/c\"], chromosome=\"2\")\n",
    "\n",
    "print(\"feature 2/000/010\")\n",
    "print(np.load(\"temp/feature/2/000/010.npy\")[:, 0])\n",
    "print(np.load(\"temp/feature/2/000/010.npy\")[:, 1])\n",
    "print(\"sample a/2/000/010\")\n",
    "print(np.load(\"temp/embd/a/2/000/010.npy\")[:, 768])\n",
    "print(\"sample b/2/000/010\")\n",
    "print(np.load(\"temp/embd/b/2/000/010.npy\")[:, 768])\n",
    "print(\"sample c/2/000/010\")\n",
    "print(np.load(\"temp/embd/c/2/000/010.npy\")[:, 768])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(70, 3)\n",
      "hash_idx, bucket_idx, distance\n",
      "[[10.         2.         2.715723 ]\n",
      " [10.         3.         0.       ]\n",
      " [10.         4.         2.634376 ]\n",
      " [10.         5.         2.7557654]\n",
      " [10.         6.         0.       ]\n",
      " [10.         7.         2.5555692]\n",
      " [11.         1.         0.       ]\n",
      " [11.         2.         1.9527494]\n",
      " [11.         3.         2.190697 ]\n",
      " [11.         4.         2.7221084]]\n",
      "after select top 15% of feature\n",
      "hash_idx, bucket_idx, distance\n",
      "[[12.         3.         3.570308 ]\n",
      " [12.         4.         3.0114646]\n",
      " [14.         0.         3.5584853]\n",
      " [14.         1.         4.5021667]\n",
      " [15.         8.         3.4176254]\n",
      " [21.         2.         2.975521 ]\n",
      " [22.         8.         3.1255062]\n",
      " [26.         1.         3.6105146]\n",
      " [32.         6.         3.7654283]\n",
      " [34.         9.         4.198085 ]]\n"
     ]
    }
   ],
   "source": [
    "## getFeature { chromosome : (hash_idx, bucket_idx, distance) }\n",
    "\n",
    "from src.embd2repre import Selector\n",
    "\n",
    "selector = Selector(\"temp/feature\")\n",
    "feature_c = selector.getFeature(chromosome=\"2\")[\"2\"]\n",
    "\n",
    "print(feature_c.shape)\n",
    "print(\"hash_idx, bucket_idx, distance\")\n",
    "print(feature_c[:10])\n",
    "\n",
    "top_k = 0.15\n",
    "top_k = int(len(feature_c) * top_k)\n",
    "order = np.sort(np.argsort(feature_c[:, 2])[-top_k:])\n",
    "feature_c = feature_c[order]\n",
    "\n",
    "print(\"after select top 15% of feature\")\n",
    "print(\"hash_idx, bucket_idx, distance\")\n",
    "print(feature_c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "applyFeature/0/2 :   0%|          | 0/8 [00:00<?, ?hash/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "applyFeature/0/2 : 100%|██████████| 8/8 [00:00<00:00, 355.00hash/s]\n",
      "applyFeature/1/2 : 100%|██████████| 8/8 [00:00<00:00, 246.81hash/s]\n",
      "applyFeature/2/2 : 100%|██████████| 8/8 [00:00<00:00, 157.31hash/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[12363. 12433. 14036. 14158. 15855.    nan    nan    nan    nan    nan]\n",
      "(10, 770)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "## applyFeature (:768 embd, 768 pos, 769 embd_idx)\n",
    "\n",
    "from src.embd2repre import Selector\n",
    "\n",
    "selector = Selector(\"temp/feature\")\n",
    "selector.applyFeature([\"temp/embd/a\", \"temp/embd/b\", \"temp/embd/c\"], chromosome=\"2\")\n",
    "\n",
    "feature = np.load(\"temp/embd/a/2/feature.npy\")\n",
    "print(feature[:, 768])\n",
    "print(feature.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read profile\n",
    "profile = pd.read_csv(\"data/profile.csv\")\n",
    "profile = profile[profile[\"dataset\"] == \"tcgaskcm\"]\n",
    "profile = pd.concat([\n",
    "    profile[profile[\"easy\"] == 1], profile[profile[\"hard\"] == 1]\n",
    "]).reset_index(drop=True)\n",
    "# random sample train, valid, test using pandas sample\n",
    "profile_train = profile.sample(frac=0.6, random_state=0)\n",
    "profile_valid = profile.drop(profile_train.index).sample(frac=0.5, random_state=0)\n",
    "profile_test  = profile.drop(profile_train.index).drop(profile_valid.index)\n",
    "\n",
    "#profile_train = profile[profile[\"train\"] == 1]\n",
    "#profile_valid = profile[profile[\"valid\"] == 1]\n",
    "#profile_test  = profile[profile[\"test\"]  == 1]\n",
    "# representation\n",
    "repre_train = np.hstack([np.load(row[\"repre_path\"]) for _, row in profile_train.iterrows()]).T  # (497, 768)\n",
    "repre_valid = np.hstack([np.load(row[\"repre_path\"]) for _, row in profile_valid.iterrows()]).T  # (216, 768)\n",
    "repre_test  = np.hstack([np.load(row[\"repre_path\"]) for _, row in profile_test.iterrows() ]).T  # (215, 768)\n",
    "repre = np.vstack([repre_train, repre_valid, repre_test])   # (928, 768)\n",
    "# PCA\n",
    "pca = sklearn.decomposition.PCA()\n",
    "pca = pca.fit(repre)\n",
    "repre_train = pca.transform(repre_train)    # (497, 768)\n",
    "repre_valid = pca.transform(repre_valid)    # (216, 768)\n",
    "repre_test  = pca.transform(repre_test)     # (215, 768)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.scatter(repre_train[:, 0], repre_train[:, 1], s=5)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "start, end = 0, 100\n",
    "t1, t2 = 0.4, 0.6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## T Cells CD8\n",
    "# label\n",
    "label_train = profile_train[\"Leukocyte Fraction\"]\n",
    "label_valid = profile_valid[\"Leukocyte Fraction\"]\n",
    "label_test  = profile_test [\"Leukocyte Fraction\"]\n",
    "# prediction\n",
    "model = sklearn.linear_model.LinearRegression().fit(repre_train[:, start:end], label_train)\n",
    "#model = sklearn.svm.SVR().fit(repre_train[:, start:end], label_train)\n",
    "pred_train = model.predict(repre_train[:, start:end])\n",
    "pred_valid = model.predict(repre_valid[:, start:end])\n",
    "pred_test  = model.predict(repre_test[:, start:end])\n",
    "# performance\n",
    "mae_train = sklearn.metrics.mean_absolute_error(label_train, pred_train)\n",
    "mae_valid = sklearn.metrics.mean_absolute_error(label_valid, pred_valid)\n",
    "print(f\"train mae: {mae_train:.3f}\")\n",
    "print(f\"valid mae: {mae_valid:.3f}\")\n",
    "# view T Cells CD8 < 0.1 as low, 0.1 <= T Cells CD8 < 0.3 as middle, \n",
    "# 0.3 <= T Cells CD8 as high, check pred_valid and profile_valid[\"T Cells CD8\"]\n",
    "# performance in classification way\n",
    "label_class_train = np.digitize(label_train, [t1, t2])\n",
    "label_class_valid = np.digitize(label_valid, [t1, t2])\n",
    "pred_class_train = np.digitize(pred_train, [t1, t2])\n",
    "pred_class_valid = np.digitize(pred_valid, [t1, t2])\n",
    "# 3 * 3 confusion matrix\n",
    "confusion_matrix = sklearn.metrics.confusion_matrix(label_class_train, pred_class_train)\n",
    "print(confusion_matrix)\n",
    "confusion_matrix = sklearn.metrics.confusion_matrix(label_class_valid, pred_class_valid)\n",
    "print(confusion_matrix)\n",
    "# print accuracy\n",
    "accuracy_train = sklearn.metrics.accuracy_score(label_class_train, pred_class_train)\n",
    "accuracy_valid = sklearn.metrics.accuracy_score(label_class_valid, pred_class_valid)\n",
    "print(f\"train accuracy: {accuracy_train:.3f}\")\n",
    "print(f\"valid accuracy: {accuracy_valid:.3f}\")\n",
    "\n",
    "# plot pred_valid as x and label_valid as y\n",
    "plt.scatter(pred_train, label_train, color=\"blue\", s=10)\n",
    "plt.scatter(pred_valid, label_valid, color=\"red\", s=10)\n",
    "plt.plot([0, 0.5], [0, 0.5], color=\"black\")\n",
    "plt.xlabel(\"pred\")\n",
    "plt.ylabel(\"label\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: prediction using Th2 Cells as lable and svm as model\n",
    "svm = sklearn.svm.SVR().fit(repre_train, profile_train[\"Th2 Cells\"])\n",
    "pred_train = svm.predict(repre_train)\n",
    "pred_valid = svm.predict(repre_valid)\n",
    "mae_train = sklearn.metrics.mean_absolute_error(profile_train[\"Th2 Cells\"], pred_train)\n",
    "mae_valid = sklearn.metrics.mean_absolute_error(profile_valid[\"Th2 Cells\"], pred_valid)\n",
    "print(f\"train mae: {mae_train:.2f}\")\n",
    "print(f\"valid mae: {mae_valid:.2f}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Visulization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# profile 0: nan to 0\n",
    "# profile 1: remove nan row\n",
    "\n",
    "# representation\n",
    "profile = pd.read_csv(\"data/profile0.csv\")\n",
    "profile = profile.iloc[23:].reset_index(drop=True)\n",
    "repre = np.hstack([\n",
    "    np.load(profile.loc[i, \"repre_path\"]) for i in range(len(profile))\n",
    "    if not pd.isna(profile.loc[i, \"repre_path\"])\n",
    "]).T\n",
    "# for each column, if there are np.nan in that column, drop that column\n",
    "repre = repre[:, np.all(~np.isnan(repre), axis=0)]  # comment for 0, un for 1\n",
    "repre = np.nan_to_num(repre, nan=0.0)\n",
    "# fit\n",
    "visu_pca = sklearn.decomposition.PCA(n_components=8).fit(repre)\n",
    "visu_umap = umap.UMAP(n_components=3).fit(repre)\n",
    "# transform\n",
    "idx = [\n",
    "    i for i in range(len(profile)) if not pd.isna(profile.loc[i, \"repre_path\"])\n",
    "]\n",
    "profile.loc[idx, [f\"pca{p}\" for p in range(8)]] = visu_pca.transform(repre)\n",
    "profile.loc[idx, [f\"umap{u}\" for u in range(3)]] = visu_umap.transform(repre)\n",
    "# save\n",
    "profile.to_csv(\"data/profile2.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Profile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# { chromosome : (hash_idx, bucket_idx, distance) }\n",
    "feature = Selector(\"data/feature\").getFeature()\n",
    "# remove distance and ndarray dtype from float to int\n",
    "# { chromosome : (hash_idx, bucket_idx) }\n",
    "feature = {c: np.array([f[:2] for f in feature[c]], dtype=int) for c in feature}\n",
    "result = []\n",
    "for c in chromosome_list:\n",
    "    for hash_idx, bucket_idx in feature[c]:\n",
    "        result.append((c, hash_idx*1000 + bucket_idx*100, hash_idx*1000 + (bucket_idx+1)*100))\n",
    "result = pd.DataFrame(result, columns=[\"chromosome\", \"start\", \"end\"])\n",
    "result.to_csv(\"data/temp/feature.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hash_idx_max = {\n",
    "    \"1\": 249000,  \"2\": 243000,  \"3\": 199000,  \"4\": 191000,\n",
    "    \"5\": 182000,  \"6\": 171000,  \"7\": 160000,  \"8\": 146000,\n",
    "    \"9\": 139000, \"10\": 134000, \"11\": 136000, \"12\": 134000,\n",
    "    \"13\": 115000, \"14\": 108000, \"15\": 102000, \"16\":  91000,\n",
    "    \"17\":  84000, \"18\":  81000, \"19\":  59000, \"20\":  65000,\n",
    "    \"21\":  47000, \"22\":  51000,  \"X\": 157000,\n",
    "}\n",
    "\n",
    "## bed\n",
    "# read bed file\n",
    "bed = pd.read_csv(\n",
    "    \"data/temp/rgc_gxs_v1_hg38.bed\", \n",
    "    sep=\"\\t\", header=0, names=[\"chromosome\", \"start\", \"end\"]\n",
    ")\n",
    "# remove chr prefix\n",
    "bed[\"chromosome\"] = bed[\"chromosome\"].apply(lambda x: x[3:])\n",
    "# keep chromosome 1-22 and X\n",
    "bed = bed[bed['chromosome'].isin(chromosome_list)]\n",
    "# transform bed from range list to one hot\n",
    "bed_one_hot = {\n",
    "    c: np.zeros((hash_idx_max[c] * 1000), dtype=bool) for c in chromosome_list\n",
    "}\n",
    "for _, row in bed.iterrows():\n",
    "    chromosome, start, end = row[['chromosome', 'start', 'end']]\n",
    "    bed_one_hot[chromosome][start:end] = True\n",
    "\n",
    "## feature\n",
    "# { chromosome : (hash_idx, bucket_idx, distance) }\n",
    "feature = Selector(\"data/feature\").getFeature()\n",
    "# remove distance and ndarray dtype from float to int\n",
    "# { chromosome : (hash_idx, bucket_idx) }\n",
    "feature = {c: np.array([f[:2] for f in feature[c]], dtype=int) for c in feature}\n",
    "# transform feature from bucket list to one hot\n",
    "feature_one_hot = {\n",
    "    c: np.zeros((hash_idx_max[c] * 1000), dtype=bool) for c in chromosome_list\n",
    "}\n",
    "for c in feature:\n",
    "    for hash_idx, bucket_idx in feature[c]:\n",
    "        feature_one_hot[c][\n",
    "            hash_idx*1000 + bucket_idx*100 : hash_idx*1000 + (bucket_idx+1)*100\n",
    "        ] = True\n",
    "\n",
    "# calculate overlap\n",
    "for c in bed_one_hot:\n",
    "    overlap = np.logical_and(bed_one_hot[c], feature_one_hot[c])\n",
    "    overlap_bed = (np.sum(overlap) / np.sum(bed_one_hot[c])) * 100 \n",
    "    overlap_feature = (np.sum(overlap) / np.sum(feature_one_hot[c])) * 100\n",
    "    print(\n",
    "        f\"Chromosome {c}:\\t\",\n",
    "        f\"overlap/bed: {np.sum(overlap):7,} / {np.sum(bed_one_hot[c]):9,} = {overlap_bed:5.2f}%\\t\",\n",
    "        f\"overlap/feature: {np.sum(overlap):7,} / {np.sum(feature_one_hot[c]):9,} = {overlap_feature:5.2f}%\"\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p_val = 1e-3\n",
    "\n",
    "snps = pd.read_csv(\"data/snps.csv\", usecols=[\"Chr\", \"Pos\", \"Pval\"])\n",
    "# filter by pval < 1e-3\n",
    "snps = snps[snps[\"Pval\"] < p_val][[\"Chr\", \"Pos\"]]\n",
    "# sort by Chr and then Pos\n",
    "snps = snps.sort_values([\"Chr\", \"Pos\"])\n",
    "# transform chromosome from int 1-23 to str 1-22, X\n",
    "snps[\"Chr\"] = snps[\"Chr\"].apply(lambda x: str(x) if x < 23 else \"X\")\n",
    "# split by chromosome, { chromosome : (pos) }\n",
    "snps = snps.groupby(\"Chr\")[\"Pos\"].apply(list).to_dict()\n",
    "\n",
    "for c, pos in snps.items():\n",
    "    snps_in_bed, snps_in_feature = 0, 0\n",
    "    for p in pos:\n",
    "        if bed_one_hot[c][p]: snps_in_bed += 1\n",
    "        if feature_one_hot[c][p]: snps_in_feature += 1\n",
    "    print(\n",
    "        f\"Chromosome {c}:\\t\",\n",
    "        f\"snps_in_bed/bed: {snps_in_bed:6,} / {len(pos):6,} = {(snps_in_bed / len(pos)) * 100:5.2f}%\\t\",\n",
    "        f\"snps_in_feature/feature: {snps_in_feature:6,} / {len(pos):6,} = {(snps_in_feature / len(pos)) * 100:5.2f}%\"\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p_val = 1e-2\n",
    "\n",
    "snps = pd.read_csv(\"data/snps.csv\", usecols=[\"Chr\", \"Pos\", \"Pval\"])\n",
    "# filter by pval < 1e-2\n",
    "snps = snps[snps[\"Pval\"] < p_val][[\"Chr\", \"Pos\"]]\n",
    "# sort by Chr and then Pos\n",
    "snps = snps.sort_values([\"Chr\", \"Pos\"])\n",
    "# transform chromosome from int 1-23 to str 1-22, X\n",
    "snps[\"Chr\"] = snps[\"Chr\"].apply(lambda x: str(x) if x < 23 else \"X\")\n",
    "# split by chromosome, { chromosome : (pos) }\n",
    "snps = snps.groupby(\"Chr\")[\"Pos\"].apply(list).to_dict()\n",
    "\n",
    "for c, pos in snps.items():\n",
    "    snps_in_bed, snps_in_feature = 0, 0\n",
    "    for p in pos:\n",
    "        if bed_one_hot[c][p]: snps_in_bed += 1\n",
    "        if feature_one_hot[c][p]: snps_in_feature += 1\n",
    "    print(\n",
    "        f\"Chromosome {c}:\\t\",\n",
    "        f\"snps_in_bed/bed: {snps_in_bed:7,} / {len(pos):7,} = {(snps_in_bed / len(pos)) * 100:5.2f}%\\t\",\n",
    "        f\"snps_in_feature/feature: {snps_in_feature:7,} / {len(pos):7,} = {(snps_in_feature / len(pos)) * 100:5.2f}%\"\n",
    "    )"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "TCGA-Onc",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
