{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import torch\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy.stats\n",
    "\n",
    "import os\n",
    "import seaborn as sns\n",
    "import tqdm.notebook as tqdm\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches\n",
    "\n",
    "\"\"\" Profile \"\"\"\n",
    "\n",
    "# dataset\n",
    "profile = pd.read_csv(\"data/stanford/profile.txt\")\n",
    "# Isolate string (i.e. su001) -> int (i.e. 1)\n",
    "profile[\"Isolate\"] = profile[\"Isolate\"].apply(lambda x: int(x[2:]))\n",
    "# Treatment pre/post -> 0/1\n",
    "profile[\"Treatment\"] = profile[\"Treatment\"].apply(lambda x: int(not \"pre\" in x))\n",
    "# Sort by Isolate (1 to 8), Treatment (pre to post), and Tissue (normal to BCC)\n",
    "profile = profile.sort_values(\n",
    "    by=[\"Isolate\", \"Treatment\", \"Tissue\"], ascending=[True, True, False]\n",
    ").reset_index(drop=True)\n",
    "# only keep run, isolate, treatment, tissue\n",
    "profile = profile[[\"Run\", \"Isolate\", \"Treatment\", \"Tissue\"]]\n",
    "\n",
    "\"\"\" Functions \"\"\"\n",
    "\n",
    "def euclideanDistance(x1: np.ndarray, x2: np.ndarray) -> np.ndarray:\n",
    "    x1norm = np.sum(x1**2, axis=1, keepdims=True)           # (N1,  1)\n",
    "    x2norm = np.sum(x2**2, axis=1, keepdims=True).T         # ( 1, N2)\n",
    "    x1x2   = 2 * np.dot(x1, x2.T)                           # (N1, N2)\n",
    "    return np.sqrt(np.maximum(0, x1norm + x2norm - x1x2))   # (N1, N2)\n",
    "\n",
    "def pearsonCorrelation(x1: np.ndarray, x2: np.ndarray) -> np.ndarray:\n",
    "    x1centered = x1 - x1.mean(axis=1, keepdims=True)        # (N1, 768)\n",
    "    x2centered = x2 - x2.mean(axis=1, keepdims=True)        # (N2, 768)\n",
    "    numerator = np.dot(x1centered, x2centered.T)            # (N1, N2)\n",
    "    x1var = np.sum(x1centered**2, axis=1, keepdims=True)    # (N1,  1)\n",
    "    x2var = np.sum(x2centered**2, axis=1, keepdims=True)    # (N2,  1)\n",
    "    denominator = np.sqrt(np.dot(x1var, x2var.T))           # (N1, N2)\n",
    "    return numerator / denominator                          # (N1, N2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" Embedding and UAMP \"\"\"\n",
    "\n",
    "method = \"finetune\"  # pretrain or finetune\n",
    "isolates = [3, 7]\n",
    "chromosome = \"6\"\n",
    "pval_thresh = 1e-5\n",
    "\n",
    "runs = []\n",
    "for isolate in isolates:\n",
    "    runs += profile[profile[\"Isolate\"] == isolate][\"Run\"].to_list()\n",
    "\n",
    "embd, umap = {}, {}\n",
    "for run in runs:\n",
    "    # load the embedding\n",
    "    embd[run] = np.load(f\"data/stanford/embd-{method}/{run}/{chromosome}.npy\")\n",
    "    # filter by p-value\n",
    "    embd[run] = embd[run][embd[run][:, 769-int(np.log10(pval_thresh))]>=1, :]\n",
    "    # only keep the first 768 columns, embedding\n",
    "    embd[run] = embd[run][:, :768]\n",
    "    # read umap of given sample id and chromosome\n",
    "    umap[run] = np.load(f\"data/stanford/umap-{method}/{run}/{chromosome}.npy\")\n",
    "    # filter reads that cover at least one variants with p-value<=pval_thresh\n",
    "    umap[run] = umap[run][umap[run][:, 3-int(np.log10(pval_thresh))]>=1, :]\n",
    "    # only keep the first 2 columns as umap coordinates\n",
    "    umap[run] = umap[run][:, :2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" Plot Distribution of Euclidean Distance \"\"\"\n",
    "\n",
    "x1embd, x2embd = embd[runs[0]], embd[runs[1]]   # (N1, 768), (N2, 768)\n",
    "distance_matrix = euclideanDistance(x1embd, x2embd)     # (N1, N2)\n",
    "\n",
    "fig, axs = plt.subplots(1, 2, figsize=(10, 5), sharex=True, sharey=True)\n",
    "for i in range(2):\n",
    "    axs[i].hist(np.max(distance_matrix, axis=int(not i)), bins=100)\n",
    "    # title\n",
    "    id, isolate, treatment, tissue = profile[profile[\"Run\"] == runs[i]].iloc[0]\n",
    "    treatment = \"Pre\" if treatment == 0 else \"Post\"\n",
    "    tissue = \"Normal\" if \"normal\" in tissue else \"Tumor\"\n",
    "    info = f\"{id} - {isolate} - {treatment} - {tissue}\"\n",
    "    axs[i].set_title(f\"{info}\", fontweight='bold', fontsize=18)\n",
    "fig.supxlabel(\"Euclidean Distance\", fontweight='bold', y=0.004, fontsize=18)\n",
    "fig.supylabel(\"Number of Reads\", fontweight='bold', x=0.008, fontsize=18)\n",
    "fig.tight_layout()\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" Plot UMAP \"\"\"\n",
    "\n",
    "top_k_reads = 4000\n",
    "\n",
    "gridsize = 100\n",
    "vmin, vmax = 2, 50\n",
    "fontsize = 18\n",
    "\n",
    "fig, axs = plt.subplots(6, 6, figsize=(30, 30), sharex=True, sharey=True)\n",
    "for i in range(len(runs)):\n",
    "    # set legend of samples\n",
    "    id, isolate, treatment, tissue = profile[profile[\"Run\"] == runs[i]].iloc[0]\n",
    "    treatment = \"Pre\" if treatment == 0 else \"Post\"\n",
    "    tissue = \"Normal\" if \"normal\" in tissue else \"Tumor\"\n",
    "    info = f\"{id} - {isolate} - {treatment} - {tissue}\"\n",
    "    axs[0,  i].set_xlabel(info, fontsize=fontsize, fontweight='bold')\n",
    "    axs[0,  i].xaxis.set_label_position(\"top\")\n",
    "    axs[i, -1].set_ylabel(info, fontsize=fontsize, fontweight='bold')\n",
    "    axs[i, -1].yaxis.set_label_position(\"right\")\n",
    "    # plot umpa of sample i before filter in diagonal\n",
    "    axs[i, i].hexbin(\n",
    "        umap[runs[i]][:, 0], umap[runs[i]][:, 1], \n",
    "        cmap=\"Reds\", gridsize=gridsize, vmin=vmin, vmax=vmax,\n",
    "    )\n",
    "    axs[i, i].set_aspect(\"equal\")\n",
    "    axs[i, i].set_xlim(-22, 22)\n",
    "    axs[i, i].set_ylim(-22, 22)\n",
    "    # plot umap of (i, :) and (:, i) after filter by distance (i, j)\n",
    "    for j in range(i+1, len(runs)):\n",
    "        x1embd, x2embd = embd[runs[i]], embd[runs[j]]   # (N1, 768), (N2, 768)\n",
    "        distance_matrix = euclideanDistance(x1embd, x2embd)     # (N1, N2)\n",
    "        x1thresh = np.sort(np.max(distance_matrix, axis=1))[-top_k_reads]\n",
    "        x2thresh = np.sort(np.max(distance_matrix, axis=0))[-top_k_reads]\n",
    "        x1filter = np.max(distance_matrix, axis=1) >= x1thresh  # (N1, )\n",
    "        x2filter = np.max(distance_matrix, axis=0) >= x2thresh  # (N2, )\n",
    "        # umap of (i, :)\n",
    "        axs[i, j].hexbin(\n",
    "            umap[runs[i]][x1filter][:, 0], umap[runs[i]][x1filter][:, 1], \n",
    "            cmap=\"Reds\", gridsize=gridsize, vmin=vmin, vmax=vmax,\n",
    "        )\n",
    "        axs[i, j].set_aspect(\"equal\")\n",
    "        axs[i, j].set_xlim(-22, 22)\n",
    "        axs[i, j].set_ylim(-22, 22)\n",
    "        # umap of (:, i)\n",
    "        axs[j, i].hexbin(\n",
    "            umap[runs[j]][x2filter][:, 0], umap[runs[j]][x2filter][:, 1], \n",
    "            cmap=\"Reds\", gridsize=gridsize, vmin=vmin, vmax=vmax,\n",
    "        )\n",
    "        axs[j, i].set_aspect(\"equal\")\n",
    "        axs[j, i].set_xlim(-22, 22)\n",
    "        axs[j, i].set_ylim(-22, 22)\n",
    "fig.supxlabel(\"UMAP1\", fontsize=fontsize, fontweight='bold', y=0.004)\n",
    "fig.supylabel(\"UMAP2\", fontsize=fontsize, fontweight='bold', x=0.008)\n",
    "fig.tight_layout()\n",
    "fig.savefig(\"temp.png\", dpi=500)\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" hexbin map for all sample \"\"\"\n",
    "\n",
    "fig, axs = plt.subplots(\n",
    "    4, 6, figsize=(30, 20), sharex=True, sharey=True, dpi=500\n",
    ")\n",
    "index = -1\n",
    "for i, ax in enumerate(axs.flat):\n",
    "    if i == 11:\n",
    "        ax.axis(\"off\")\n",
    "        continue\n",
    "    index += 1\n",
    "\n",
    "    ax.hexbin(\n",
    "        embedding_umap[profile[\"Run\"][index]][:, 0], \n",
    "        embedding_umap[profile[\"Run\"][index]][:, 1], \n",
    "        gridsize=150,\n",
    "        cmap=\"Reds\",\n",
    "        vmin=2, vmax=80,\n",
    "    )\n",
    "\n",
    "    for spine in ax.spines.values():\n",
    "        spine.set_color(\"r\" if profile[\"Treatment\"][index] == 0 else \"b\")\n",
    "        spine.set_linewidth(2)\n",
    "\n",
    "    ax.set_aspect(\"equal\")\n",
    "    ax.set_xlim(-16, 16)\n",
    "    ax.set_ylim(-16, 16)\n",
    "    ax.set_title(\n",
    "        \"Run: {}; \".format(profile[\"Run\"][index]) + \n",
    "        \"Isolate: {};\".format(profile[\"Isolate\"][index])\n",
    "    )\n",
    "fig.legend(\n",
    "    handles=[\n",
    "        matplotlib.patches.Patch(color=\"r\", label=\"pre \"), \n",
    "        matplotlib.patches.Patch(color=\"b\", label=\"post\")\n",
    "    ], loc=\"upper right\", ncol=2, fontsize=14,\n",
    ")\n",
    "fig.suptitle(\n",
    "    f\"Stanford Data; chromosome {chr}; SNPs p-val threshold (<=) {pval_thresh}\", \n",
    "    fontweight='bold', y=0.992, fontsize=16\n",
    ")\n",
    "fig.supxlabel(\"UMAP1\", fontweight='bold', y=0.005, fontsize=16)\n",
    "fig.supylabel(\"UMAP2\", fontweight='bold', x=0.010, fontsize=16)\n",
    "fig.tight_layout()\n",
    "fig.savefig(\n",
    "    os.path.join(embedding_umap_fold, \"hexbin.png\"), \n",
    "    dpi=500,\n",
    ")\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" two 1D distribution map \"\"\"\n",
    "\n",
    "x = np.linspace(-30, 30, 500)\n",
    "\n",
    "for u in [0, 1]:\n",
    "    base = scipy.stats.gaussian_kde(\n",
    "        np.concatenate([embedding_umap[run][:, u] for run in profile[\"Run\"]])\n",
    "    )(x)\n",
    "    fig, axs = plt.subplots(1, 8, figsize=(24, 6), sharex=True, sharey=True, dpi=500)\n",
    "    for i in profile[\"Isolate\"].unique():\n",
    "        ax = axs.flat[i-1]\n",
    "        for run in profile[profile[\"Isolate\"] == i][\"Run\"]:\n",
    "            density = scipy.stats.gaussian_kde(embedding_umap[run][:, u])(x)\n",
    "            ax.plot(\n",
    "                x, density - base, \n",
    "                c='r' if profile[profile[\"Run\"]==run][\"Treatment\"].values == 0 else 'b'\n",
    "            )\n",
    "            ax.set_xlim(-25, 25)\n",
    "            ax.set_ylim(-0.008, 0.008)\n",
    "            ax.set_title(f\"Isolate: {i}\")\n",
    "    fig.legend(\n",
    "        handles=[\n",
    "            matplotlib.patches.Patch(color=\"r\", label=\"pre \"), \n",
    "            matplotlib.patches.Patch(color=\"b\", label=\"post\")\n",
    "        ], loc=\"upper right\", ncol=2, fontsize=14,\n",
    "    )\n",
    "    fig.suptitle(\n",
    "        f\"Stanford Data; chromosome {chr}; SNPs p-val threshold (<=) {pval_thresh}\", \n",
    "        fontweight='bold', y=0.992, fontsize=16\n",
    "    )\n",
    "    fig.supxlabel(f\"UMAP{u+1}\", fontweight='bold', y=0.005, fontsize=16)\n",
    "    fig.supylabel(\"Density Different\", fontweight='bold', x=0.010, fontsize=16)\n",
    "    fig.tight_layout()\n",
    "    fig.savefig(\n",
    "        os.path.join(embedding_umap_fold, f\"distribution{u+1}.png\"), \n",
    "        dpi=500,\n",
    "    )\n",
    "    plt.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "TCGA-Onc",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
